Fix IRC regex handling of slashes in host

shymega opened the pull request:
As per the issue reported in #59, this commit fixes that issue with a
regex patch suggested by @vktec.

This commit has been tested with cloaks, and _appears_ to work, which is
a good improvement.

Because of the improvement, this fixes #59.

bitwalker commented on lib/exirc/utils.ex:
```diff
@@ -25,7 +25,7 @@ defmodule ExIrc.Utils do
   end
 
-  @prefix_pattern ~r/^(?<nick>[^!]+)(?:!(?:(?<user>[^@ ]+)@)?(?:(?<host>[\w.:-]+)))?$/
+  @prefix_pattern ~r/^(?<nick>[^!]+)(?:!(?:(?<user>[^@ ]+)@)?(?:(?<host>\S+[\w.:-]+)))?$/
```
Shouldn't it instead change `[\w.:-]+` to just `\S+`?

vktec replied:
Yeah, that's correct. The current version won't match, for example, `foo/`, and is also slightly less efficient.

In fact, there's also a few redundant non-capturing groups in there.

The whole line should look like the following:

```elixir
  @prefix_pattern ~r/^(?<nick>[^!]+)(?:!(?<user>[^@ ]+@)?(?<host>\S+))?$/
```

vktec replied:
Actually, the nickname matching can be improved too, since it currently accepts whitespace, and you probably shouldn't accept tabs and newlines in usernames either :P

```elixir
  @prefix_pattern ~r/^(?<nick>[^!\s]+)(?:!(?<user>[^@\s]+@)?(?<host>\S+))?$/
```

shymega replied:
@vktec Tested that pattern, get the error again.

vktec replied:
How strange. Could you try variations of the 3 modifications I've made and let me know which ones work and which don't? (I've not got Elixir installed, so can't test)

shymega replied:
On which ones? The latest patch, or the one before it?

vktec replied:
From our conversations on IRC, the one without nick or user improvements works, so try adding in the improved nick matching, then if that works try adding user matching as well. If not, disable it and try user matching.

That way, I can figure out where the issue(s) is/are.

shymega commented:
@bitwalker @vktec I've made changes to the regex, using @vktec's more efficient version. Tested on myself and @vktec's IRC server, and it _appears_ to work with cloaks with slashes in them.

shymega commented:
However, _that said_, Travis is reporting build failures. These should be addressed ASAP too..

bitwalker commented:
Could we add some tests not just for the cloaks, but for standard nicks/hosts as well? I'm sure it works, but I'm leery of breaking changes, so I prefer to have it encoded in tests that everything is still good to go. Once that's in place, I'll get this merged and published. Thanks!

bitwalker commented:
Yeah those are failed tests which are impacted by this change, it looks like either the regex isn't exactly right, or you need to update the code consuming the pattern to make sure the new groupings are handled properly.

shymega commented:
Hello,

Paul Schoenfelder <notifications@github.com> writes:

> Could we add some tests not just for the cloaks, but for standard
> nicks/hosts as well? I'm sure it works, but I'm leery of breaking
> changes, so I prefer to have it encoded in tests that everything is
> still good to go. Once that's in place, I'll get this merged and
> published. Thanks!

So am I ref: breaking changes. We can put some tests in, sure. I'll do
that later, once I get back to my desk.

-- 
Kind regards,
Dom Rodriguez (shymega/dzr{_,})

"Freedom is the freedom to say two plus two make four.
 If that is granted, all else follows."
	-- George Orwell, Nineteen Eighty-Four.

shymega commented:
Hello,

Paul Schoenfelder <notifications@github.com> writes:

> Yeah those are failed tests which are impacted by this change, it
> looks like either the regex isn't exactly right, or you need to
> update the code consuming the pattern to make sure the new groupings
> are handled properly.

Looks like the regex isn't right. Check this out from some of my
debugging:

```
iex(10)> data = ':nick!user.name@irc.example.org PART #channel'
':nick!user.name@irc.example.org PART #channel'
iex(11)> message = ExIrc.Utils.parse(data)                     
%IrcMessage{args: ["#channel"], cmd: "PART", ctcp: false,
 host: "irc.example.org", nick: "nick", server: [], user: "user.name@"}
iex(12)> 
```

So clearly, the regex is having some issues. I'm not great shakes at
regexs, perhaps you or @vktec could lend a hand in debuggin'?

Cheers!

-- 
Kind regards,
Dom Rodriguez (shymega/dzr{_,})

"Freedom is the freedom to say two plus two make four.
 If that is granted, all else follows."
	-- George Orwell, Nineteen Eighty-Four.

bitwalker commented:
With the following regex:

```elixir
 ~r/^(?<nick>[^!\s]+)(?:!(?:(?<user>[^@\s]+)@)?(?:(?<host>[\S]+)))?$/
```

It works fine, for both cloaked/uncloaked. It's the removal of some of the groupings that is throwing things off. I would suggest using the above pattern in this PR.

shymega commented:
Added the new regex, and added tests for cloaked/uncloaked users. Passes locally, waiting for Travis.

Once Travis passes, @bitwalker - feel free to merge in :smile:

bitwalker commented:
Thanks!

vktec commented:
Oops, my bad. I put the `@` in the wrong place. Looks like it's working now though :)
